skxray/cdi.py

```diff
@@ -96,32 +99,321 @@ def gauss(dims, sigma):
     return y / np.sum(y)
 
 
-def convolution(array1, array2):
+_fft_helper = lambda x: np.fft.ifftshift(np.fft.fftn(np.fft.fftshift(x)))
+_ifft_helper = lambda x: np.fft.ifftshift(np.fft.ifftn(np.fft.fftshift(x)))
+
+
+def pi_modulus(recon_pattern,
+               diffracted_pattern,
+               offset_v=1e-12):
+    """
+    Transfer sample from real space to q space.
+    Use constraint based on diffraction pattern from experiments.
+
+    Parameters
+    ----------
+    recon_pattern : array
+        reconstructed pattern in real space
+    diffracted_pattern : array
+        diffraction pattern from experiments
+    offset_v : float
+        add small value to avoid the case of dividing something by zero
+
+    Returns
+    -------
+    arr : array
+        updated pattern in q space
+    """
+    diff_tmp = _fft_helper(recon_pattern) / np.sqrt(np.size(recon_pattern))
+    index = diffracted_pattern > 0
+    diff_tmp[index] = (diffracted_pattern[index] *
+                       diff_tmp[index] / (np.abs(diff_tmp[index]) + offset_v))
+    return _ifft_helper(diff_tmp) * np.sqrt(np.size(diffracted_pattern))
```

Comment:
I think you can do both of the multiplications here which will give a bit of a speed up (and make sure to use `()` so that the two scalar multiplications happen first).